x-pack/plugins/ml/public/settings/calendars/edit/new_calendar.js

```diff
@@ -0,0 +1,310 @@
+        selectedGroupOptions,
+        selectedCalendar
+      });
+    } catch (error) {
+      console.log(error);
+      this.setState({ loading: false });
+      toastNotifications.addDanger('An error occurred loading calendar form data. Try refreshing the page.');
+    }
+  }
+
+  onCreate = async () => {
+    const calendar = this.setUpCalendarForApi();
+    this.setState({ saving: true });
+
+    try {
+      await ml.addCalendar(calendar);
+      window.location = `${chrome.getBasePath()}/app/ml#/settings/calendars_list`;
+    } catch (error) {
+      console.log('Error saving calendar', error);
+      this.setState({ saving: false });
+      toastNotifications.addDanger(`An error occurred creating calendar ${calendar.calendarId}`);
+    }
+  }
+
+  onEdit = async () => {
+    const calendar = this.setUpCalendarForApi();
+    this.setState({ saving: true });
+
+    try {
+      await ml.updateCalendar(calendar);
+      window.location = `${chrome.getBasePath()}/app/ml#/settings/calendars_list`;
+    } catch (error) {
+      console.log('Error saving calendar', error);
+      this.setState({ saving: false });
+      toastNotifications.addDanger(`An error occurred saving calendar ${calendar.calendarId}. Try refreshing the page.`);
+    }
+  }
+
+  setUpCalendarForApi = () => {
+    const {
+      formCalendarId,
+      description,
+      events,
+      selectedGroupOptions,
+      selectedJobOptions,
+    } = this.state;
+
+    const jobIds = selectedJobOptions.map((option) => option.label);
+    const groupIds = selectedGroupOptions.map((option) => option.label);
+
+    // set up calendar
+    const calendar = {
+      calendarId: formCalendarId,
+      description,
+      events,
+      job_ids: [...jobIds, ...groupIds]
+    };
+
+    return calendar;
+  }
+
+  onCreateGroupOption = (newGroup) => {
+    const newOption = {
+      label: newGroup,
+    };
+    // Select the option.
+    this.setState(prevState => ({
+      selectedGroupOptions: prevState.selectedGroupOptions.concat(newOption),
+    }));
+  };
+
+  onJobSelection = (selectedJobOptions) => {
+    this.setState({
+      selectedJobOptions,
+    });
+  };
+
+  onGroupSelection = (selectedGroupOptions) => {
+    this.setState({
+      selectedGroupOptions,
+    });
+  };
+
+  onCalendarIdChange = (e) => {
+    const isValid = validateCalendarId(e.target.value);
+
+    this.setState({
+      formCalendarId: e.target.value,
+      isNewCalendarIdValid: isValid
+    });
+  };
+
+  onDescriptionChange = (e) => {
+    this.setState({
+      description: e.target.value,
+    });
+  };
+
+  showImportModal = () => {
+    this.setState(prevState => ({
+      isImportModalVisible: !prevState.isImportModalVisible,
+    }));
+  }
+
+  closeImportModal = () => {
+    this.setState({
+      isImportModalVisible: false,
+    });
+  }
+
+  onEventDelete = (eventId) => {
+    this.setState(prevState => ({
+      events: prevState.events.filter(event => event.event_id !== eventId)
```

Comment:
I think there is a problem with this for events which have been added, but before the calendar has been saved. Looks like events only pick up an ID after saving, so if you delete a 'new' event (as in one which has yet to be saved), all 'new' events end up being deleted.

Reply:
Heya @peteharverson  - this is fixed now. The temp event id wasn't being passed on to the new_calendar form so there was no id to keep track of deletions. 👍